### DISCUSSION TITLE
Access Codes

### DISCUSSION THREAD

**Discussion by bameyrick** [ONSdigital/design-system#1360] [Feedback]
Use this place to discuss the [access codes pattern](https://ons-design-system.netlify.com/patternes/uac) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by bameyrick** [comment]
The UAC component could do with being more configurable, as it might not always be the same length, groupings, character sets etc.

**Reply by jrbarnes9** [reply]
Resolved in #429

**Comment by bameyrick** [comment]
The full UAC code isn't visible in the input because IE11 adds a remove button into the input. This could be turned off, but I wanted to check with @will-madebymoon if we should do this, and if we do turn it off, whether we turn it off on every input?

**Comment by Paul-Joel** [comment]
I've just tried to implement the UAC element, and it does not generate a 'name' attribute, so I can't access any data from it

Calling with this:
    {{ onsUACInput({
        "id": "uac",
        "name": "uac",
        "label": {
            "text": "Enter the 16 character code printed on the letter",
            "description": "Keep this code safe and only share it with others in your household"
        },
        "securityMessage": "Your personal information is protected by law and will be kept confidential."
    }) }}

Generates this:

<p class="field uac__field">
--
  |  
  |  
  |  
  |  
  |  
  | <label
  | class="label"
  | for="uac"
  |  
  |  
  | >Enter the 16 character code printed on the letter
  | <br>
  | <span class="label__description">Keep this code safe and only share it with others in your household</span>
  | </label>
  |  
  |  
  |  
  |  
  | <input
  | type="text"
  | id="uac"
  | class="input input--text input-type__input uac__input js-uac"
  |  
  |  
  |  
  |  
  |  
  |  
  |  
  |  
  |  
  | maxlength="19.0" data-group-size="4" autocomplete="off"
  | />
  |  
  |  
  |  
  |  
  |  
  |  
  | </p>
  |  
  |  
  |  
  |  
  |  
  |  
  | <div class="uac__security u-mt-m">
  | <div class="uac__icon"><span class="icon icon--lock"></span></div>
  | <p class="uac__message">Your personal information is protected by law and will be kept confidential.</p>
  | </div>

**Comment by jrbarnes9** [comment]
UAC input entry doesn't reflect appearance of code on paper (eg. uppercase, font)

**Reply by jrbarnes9** [reply]
Fixed #696

**Comment by jrbarnes9** [comment]
Remove required attribute from Access Code macro

**Comment by Paul-Joel** [comment]
There appears to be an issue between the UAC pattern and the updated onsPanels. The UAC call now errors in RHUI, with teh following end of trace
  File "/Users/pauljoel/PycharmProjects/census-rh-ui/app/templates/start.html", line 103, in block "main"
    onsUACInput({
  File "/Users/pauljoel/.local/share/virtualenvs/census-rh-ui-3GD5Gbtn/lib/python3.6/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "/Users/pauljoel/PycharmProjects/census-rh-ui/app/templates/components/uac/_macro.njk", line 30, in template
    {% call onsPanel() %}
  File "/Users/pauljoel/.local/share/virtualenvs/census-rh-ui-3GD5Gbtn/lib/python3.6/site-packages/jinja2/runtime.py", line 679, in _invoke
    rv = self._func(*arguments)
  File "/Users/pauljoel/PycharmProjects/census-rh-ui/app/templates/components/panel/_macro.njk", line 2, in template
    {% if params.classes is defined %}
  File "/Users/pauljoel/.local/share/virtualenvs/census-rh-ui-3GD5Gbtn/lib/python3.6/site-packages/jinja2/environment.py", line 471, in getattr
    return getattr(obj, attribute)
jinja2.exceptions.UndefinedError: parameter 'params' was not provided

**Comment by jrbarnes9** [comment]
Access codes pattern error example showing a suffix number
<img width="835" alt="Screenshot 2021-01-08 at 15 49 35" src="https://user-images.githubusercontent.com/43346934/104034755-38506000-51c9-11eb-8c82-c3818b60502f.png">

**Comment by jrbarnes9** [comment]
Access codes: Duplicate strong tags in census example warning panel

Remove `<strong>` tags from Census example warning in twisty.  This is dealt with by the warning panel component.

**Reply by jrbarnes9** [reply]
Raised #1362

**Comment by jrbarnes9** [comment]
Access code styles are being overridden by input styles (`font-family` and `font-size`)

Also, the attribute `autocapitalize` is not required if `type: number`

**Reply by jrbarnes9** [reply]
Raised #1768

**Comment by jrbarnes9** [comment]
Access code input width should automatically adapt according to the `maxlength`, `groupSize` and `type` parameters

**Reply by jrbarnes9** [reply]
Raised #2127
